src/field/vecops.cxx

```diff
@@ -34,69 +34,85 @@
  * Gradient operators
  **************************************************************************/
 
-const Vector2D Grad(const Field2D &f, CELL_LOC UNUSED(outloc)) {
+const Vector2D Grad(const Field2D &f, CELL_LOC outloc) {
   Vector2D result(f.getMesh());
 
   TRACE("Grad( Field2D )");
   
-  result.x = DDX(f);
-  result.y = DDY(f);
-  result.z = DDZ(f);
+  if (outloc == CELL_DEFAULT) {
+    outloc = f.getLocation();
+  }
+
+  result.x = DDX(f, outloc);
+  result.y = DDY(f, outloc);
+  result.z = DDZ(f, outloc);
 
   result.covariant = true;
 
   return result;
 }
 
-const Vector3D Grad(const Field3D &f, 
-                    CELL_LOC outloc_x, CELL_LOC outloc_y, CELL_LOC outloc_z) {
+const Vector3D Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
+                    CELL_LOC outloc_z) {
   Vector3D result(f.getMesh());
 
   TRACE("Grad( Field3D )");
 
-  if(outloc_x == CELL_DEFAULT)
+  if (outloc_x == CELL_DEFAULT) {
     outloc_x = f.getLocation();
-  if(outloc_y == CELL_DEFAULT)
+  }
+  if (outloc_y == CELL_DEFAULT) {
     outloc_y = f.getLocation();
-  if(outloc_z == CELL_DEFAULT)
+  }
+  if (outloc_z == CELL_DEFAULT) {
     outloc_z = f.getLocation();
+  }
 
   result.x = DDX(f, outloc_x);
   result.y = DDY(f, outloc_y);
   result.z = DDZ(f, outloc_z);
 
   result.covariant = true;
-  
+
   return result;
 }
 
 const Vector3D Grad(const Field3D &f, CELL_LOC outloc) {
-  if(outloc == CELL_VSHIFT)
+  if (outloc == CELL_VSHIFT) {
     return Grad(f, CELL_XLOW, CELL_YLOW, CELL_ZLOW);
-  
+  }
+
+  if (outloc == CELL_DEFAULT) {
+    outloc = f.getLocation();
+  }
+
   return Grad(f, outloc, outloc, outloc);
 }
 
-const Vector3D Grad_perp(const Field3D &f, CELL_LOC outloc_x,
-                         CELL_LOC UNUSED(outloc_y), CELL_LOC outloc_z) {
+const Vector3D Grad_perp(const Field3D &f, CELL_LOC outloc_x, CELL_LOC UNUSED(outloc_y),
+                         CELL_LOC outloc_z) {
   Vector3D result(f.getMesh());
 
   TRACE("Grad_perp( Field3D )");
 
-  if(outloc_x == CELL_DEFAULT)
+  if (outloc_x == CELL_DEFAULT) {
     outloc_x = f.getLocation();
-  if(outloc_z == CELL_DEFAULT)
+  }
+  if (outloc_z == CELL_DEFAULT) {
     outloc_z = f.getLocation();
+  }
 
-  Coordinates* metric_x = mesh->coordinates(outloc_x);
-  Coordinates* metric_z = mesh->coordinates(outloc_z);
+  Coordinates *metric_x = mesh->coordinates(outloc_x);
+  Coordinates *metric_z = mesh->coordinates(outloc_z);
 
-  result.x = DDX(f, outloc_x) - metric_x->g_12*DDY(f, outloc_x) / SQ(metric_x->J * metric_x->Bxy);
+  result.x = DDX(f, outloc_x) -
+             metric_x->g_12 * DDY(f, outloc_x) / SQ(metric_x->J * metric_x->Bxy);
   result.y = 0.0;
```

Comment:
Note that `result.y` doesn't get a location set meaning it's centred unlike the other two fields. This feels odd.

Reply:
True. Should it use `outloc_y` similarly to the other two locations?

Reply:
Yes I think that's correct.

Reply:
Now handled by `Vector3D::setLocation`